presto-spi/src/main/java/com/facebook/presto/spi/Node.java

```diff
@@ -36,4 +37,6 @@
     boolean isResourceManager();
 
     boolean isCatalogServer();
+
+    Optional<String> getPoolType();
```

Comment:
Couple of suggestions.
1. String -> Enum for capturing types.
2. Have you though about having a `default NodePoolType` method instead of returning an `Optional<String>` so that your calling code in `SectionExecutionFactory` becomes simpler.
`return Optional.of(node -> node.getPoolType().map(poolType -> workerPoolType.equals(poolType)).orElse(true));`
->
`return Optional.of(node -> node.getPoolType().equals(workerPoolType));`

Reply:
Great point! Using enum was in my TODO bucket before I moved the changes out of failure recovery prototype, not sure how I missed this :) will make the change.